tensorflow-core/tensorflow-core-native/src/main/native/org/tensorflow/internal/c_api/gradients_impl.cc

```diff
@@ -0,0 +1,98 @@
+/* Copyright 2024 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "graph.h"
+#include "tsl/platform/errors.h"
+#include "tensorflow/c/c_api.h"
+#include "tensorflow/c/tf_status.h"
+#include "tensorflow/cc/framework/grad_op_registry.h"
+
+namespace tensorflow {
+    namespace java {
+        using namespace tsl;
+        using namespace std;
+
+        unordered_map<string, TF_GradFuncAdapter> g_grad_func_adapters;
+
+        template <typename T, typename U> T* struct_cast(U* ptr) {
+            return static_cast<T*>(static_cast<void*>(ptr));
```

Comment:
This seems like crimes are occurring. Is the type information not present in the headers?

Reply:
Haha! I actually got that from the TensorFlow source code [here](https://github.com/tensorflow/tensorflow/blob/9d637f69f699c0c422716b56153a8b27b681891a/tensorflow/c/c_api.cc#L658). Now, it does not mean that if it is in TensorFlow code that it is not evil ;) Actually it is used to carry out C++ pointers into a C struct that is left undefined in the headers, yes. That's how the C API does it too.

Reply:
Might be worth putting a comment pointing at the C API, because that cast is just nasty and I'd like to remember to blame someone else for it.